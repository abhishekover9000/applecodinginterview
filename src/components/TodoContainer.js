import React, { Component } from "react";
import TodoAddItem from "./TodoAddItem";
import TodoItem from "./TodoItem";
import * as constants from "./TodoConst";
import Button from "@material-ui/core/Button";

class TodoContainer extends Component {
  state = {
    items: [],
    isUserAddingNewItem: false,
    timeOut: null
  };

  toggleUserAdd = () => {
    this.setState({ isUserAddingNewItem: true });
  };

  addNewItem = itemName => {
    const items = this.state.items;
    items.push({
      title: itemName,
      isActive: false,
      isComplete: false,
      isSuccess: false,
      isDeleted: false,
      isDeleteInGrace: false
    });
    this.setState({ items, isUserAddingNewItem: false });
  };

  updateItem = (itemIndex, action) => {
    const item = this.state.items[itemIndex];
    switch (action) {
      case constants.ACTIVATE:
        item.isActive = true;
        break;
      case constants.DEACTIVATE:
        item.isActive = false;
        break;
      case constants.COMPLETE:
        item.isComplete = true;
        item.isActive = false;
        break;
      case constants.INCOMPLETE:
        item.isComplete = false;
        break;
      case constants.SUCCESS:
        item.isSuccess = true;
        item.isActive = false;
        break;
      case constants.NOT_SUCCESS:
        item.isSuccess = false;
        break;
      case constants.DELETE:
        item.isDeleted = true;
        item.isActive = false;
        break;
      case constants.UNDO:
        item.isDeleted = false;
        item.isDeleteInGrace = false;
        break;
      case constants.DELETE_GRACE:
        item.isDeleteInGrace = true;
        break;
    }

    this.setState({
      items: this.state.items
    });
  };

  render() {
    const { items, isUserAddingNewItem } = this.state;

    const itemsList = items.map((item, i) => {
      if (!item.isDeleted) {
        return (
          <TodoItem key={i} item={item} index={i} update={this.updateItem} />
        );
      }
    });
    return (
      <div>
        <div className="itemList">
          {items.length > 0 ? (
            <React.Fragment> {itemsList} </React.Fragment>
          ) : (
            <div> "Nothing to do yet.." </div>
          )}
        </div>
        {isUserAddingNewItem ? (
          <TodoAddItem addItem={this.addNewItem} />
        ) : (
          <Button
            className="addUserButton"
            color="primary"
            variant="contained"
            onClick={this.toggleUserAdd}
          >
            {" "}
            Add Item{" "}
          </Button>
        )}
      </div>
    );
  }
}
export default TodoContainer;
