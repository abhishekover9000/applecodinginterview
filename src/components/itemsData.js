export const itemsData = [
  {
    name: "test",
    isComplete: false,
    isSuccess: false,
    isDeleted: false,
    isDeleteInGrace: false
  },
  {
    name: "test2",
    isComplete: false,
    isSuccess: true,
    isDeleted: false,
    isDeleteInGrace: false
  }
];
