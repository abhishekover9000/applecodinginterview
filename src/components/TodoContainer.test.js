import React from "react";
import { shallow, configure } from "enzyme";
import TodoContainer from "./TodoContainer";
import { itemsData } from "./itemsData";

describe("TodoContainer", () => {
  const app = shallow(<TodoContainer />);

  it("renders properly", () => {
    expect(app).toMatchSnapshot();
  });

  it("does not display an item when item list is empty", () => {
    expect(app.find(".item").exists()).toEqual(false);
  });

  app.setState({ items: itemsData });
  it("does displays items when item list is not empty", () => {
    // set time out gives app time to update state
    setTimeout(() => {
      expect(app.find(".item").exists()).toEqual(true);
    }, 1000);
  });

  it("ToggleUserAdd changes state", () => {
    app.find(".addUserButton").simulate("click");
    expect(app.state("isUserAddingNewItem")).toEqual(true);
  });
});
