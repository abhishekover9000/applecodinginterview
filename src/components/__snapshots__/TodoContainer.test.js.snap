// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoContainer renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="itemList"
        >
          <React.Fragment>
             
            <TodoItem
              index={0}
              item={
                Object {
                  "isComplete": false,
                  "isDeleteInGrace": false,
                  "isDeleted": false,
                  "isSuccess": false,
                  "name": "test",
                }
              }
              update={[Function]}
            />
            <TodoItem
              index={1}
              item={
                Object {
                  "isComplete": false,
                  "isDeleteInGrace": false,
                  "isDeleted": false,
                  "isSuccess": true,
                  "name": "test2",
                }
              }
              update={[Function]}
            />
             
          </React.Fragment>
        </div>,
        <WithStyles(Button)
          className="addUserButton"
          color="primary"
          onClick={[Function]}
          variant="contained"
        >
           
          Add Item
           
        </WithStyles(Button)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <React.Fragment>
             
            <TodoItem
              index={0}
              item={
                Object {
                  "isComplete": false,
                  "isDeleteInGrace": false,
                  "isDeleted": false,
                  "isSuccess": false,
                  "name": "test",
                }
              }
              update={[Function]}
            />
            <TodoItem
              index={1}
              item={
                Object {
                  "isComplete": false,
                  "isDeleteInGrace": false,
                  "isDeleted": false,
                  "isSuccess": true,
                  "name": "test2",
                }
              }
              update={[Function]}
            />
             
          </React.Fragment>,
          "className": "itemList",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              " ",
              Array [
                <TodoItem
                  index={0}
                  item={
                    Object {
                      "isComplete": false,
                      "isDeleteInGrace": false,
                      "isDeleted": false,
                      "isSuccess": false,
                      "name": "test",
                    }
                  }
                  update={[Function]}
                />,
                <TodoItem
                  index={1}
                  item={
                    Object {
                      "isComplete": false,
                      "isDeleteInGrace": false,
                      "isDeleted": false,
                      "isSuccess": true,
                      "name": "test2",
                    }
                  }
                  update={[Function]}
                />,
              ],
              " ",
            ],
          },
          "ref": null,
          "rendered": Array [
            " ",
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "class",
              "props": Object {
                "index": 0,
                "item": Object {
                  "isComplete": false,
                  "isDeleteInGrace": false,
                  "isDeleted": false,
                  "isSuccess": false,
                  "name": "test",
                },
                "update": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "index": 1,
                "item": Object {
                  "isComplete": false,
                  "isDeleteInGrace": false,
                  "isDeleted": false,
                  "isSuccess": true,
                  "name": "test2",
                },
                "update": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            " ",
          ],
          "type": Symbol(react.fragment),
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            " ",
            "Add Item",
            " ",
          ],
          "className": "addUserButton",
          "color": "primary",
          "onClick": [Function],
          "variant": "contained",
        },
        "ref": null,
        "rendered": Array [
          " ",
          "Add Item",
          " ",
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="itemList"
          >
            <React.Fragment>
               
              <TodoItem
                index={0}
                item={
                  Object {
                    "isComplete": false,
                    "isDeleteInGrace": false,
                    "isDeleted": false,
                    "isSuccess": false,
                    "name": "test",
                  }
                }
                update={[Function]}
              />
              <TodoItem
                index={1}
                item={
                  Object {
                    "isComplete": false,
                    "isDeleteInGrace": false,
                    "isDeleted": false,
                    "isSuccess": true,
                    "name": "test2",
                  }
                }
                update={[Function]}
              />
               
            </React.Fragment>
          </div>,
          <WithStyles(Button)
            className="addUserButton"
            color="primary"
            onClick={[Function]}
            variant="contained"
          >
             
            Add Item
             
          </WithStyles(Button)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <React.Fragment>
               
              <TodoItem
                index={0}
                item={
                  Object {
                    "isComplete": false,
                    "isDeleteInGrace": false,
                    "isDeleted": false,
                    "isSuccess": false,
                    "name": "test",
                  }
                }
                update={[Function]}
              />
              <TodoItem
                index={1}
                item={
                  Object {
                    "isComplete": false,
                    "isDeleteInGrace": false,
                    "isDeleted": false,
                    "isSuccess": true,
                    "name": "test2",
                  }
                }
                update={[Function]}
              />
               
            </React.Fragment>,
            "className": "itemList",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                " ",
                Array [
                  <TodoItem
                    index={0}
                    item={
                      Object {
                        "isComplete": false,
                        "isDeleteInGrace": false,
                        "isDeleted": false,
                        "isSuccess": false,
                        "name": "test",
                      }
                    }
                    update={[Function]}
                  />,
                  <TodoItem
                    index={1}
                    item={
                      Object {
                        "isComplete": false,
                        "isDeleteInGrace": false,
                        "isDeleted": false,
                        "isSuccess": true,
                        "name": "test2",
                      }
                    }
                    update={[Function]}
                  />,
                ],
                " ",
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "class",
                "props": Object {
                  "index": 0,
                  "item": Object {
                    "isComplete": false,
                    "isDeleteInGrace": false,
                    "isDeleted": false,
                    "isSuccess": false,
                    "name": "test",
                  },
                  "update": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "class",
                "props": Object {
                  "index": 1,
                  "item": Object {
                    "isComplete": false,
                    "isDeleteInGrace": false,
                    "isDeleted": false,
                    "isSuccess": true,
                    "name": "test2",
                  },
                  "update": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              " ",
            ],
            "type": Symbol(react.fragment),
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              " ",
              "Add Item",
              " ",
            ],
            "className": "addUserButton",
            "color": "primary",
            "onClick": [Function],
            "variant": "contained",
          },
          "ref": null,
          "rendered": Array [
            " ",
            "Add Item",
            " ",
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
